### ISSUE TITLE
AAS web ui performance issues in large scale scenarios

### ISSUE THREAD

**Issue by JStaWZL** [eclipse-basyx/basyx-applications#87]
We currently deployed the AAS web UI (docker image version aas-gui:v2-240227) together with the eclipse basyx java V2 components for a customer who is interested in using AAS in his production. What we discovered was that the web UI works fine for tens up to a few hundred AAS instances in the backend but is not usable anymore for higher numbers. From an outside view it appears to be the case that data for each and every AAS in the backend is loaded into the Web UI and at some point this leads to a slow down and eventually a crash. We could not find any configuration parameter or something that might help in our case reading the documentation on Github. So we decided to ask if there are any plans to increase usability of the application in larger scale scenarios with thousands of AAS instances in the backend or if we overlooked something.

**Comment by aaronzi** [comment]
Hello @JStaWZL,

we are aware of this performance issue. I assume this issue does not crash the BaSyx Java Server SDK components themselfes but rather the browser tab of the AAS UI when trying to visualize this high number of shells. Please correct me if I'm wrong.

I have a feature on my development roadmap for the AAS UI which will implement pagination for the AAS List on the left side of the UI. This will drastically decrease the amount of shells that are fetched at the same time and should resolve the issue.

The only drawback is that search by idShort (Name) will only work for the shells that are part of the current page. Unfortunately, those are the only options for retrieving shells from the AAS Registry:

![image](https://github.com/eclipse-basyx/basyx-applications/assets/17433117/2328350e-f2b5-4208-9ee5-7366aef838d4)

I will try to add this to the next release as well which should take place at the end of March.

**Comment by aaronzi** [comment]
I found a good alternative that does not break the search functionality. What I will do instead of using pagination is a so-called virtual scroller. With this mechanism, only visible AAS in the AAS List will be rendered. This means that only up to let's say 20 AAS are being rendered at the same time. This should also drastically reduce the required performance while still keeping the option to search by idShort. I will release a test image on my own DockerHub later today. I would appreciate it if you could test this image for your use case because I don't have that amount of AASs to work with so it's hard for me to verify that it is working as expected.

I will add another comment to this issue when the image is online and give you the link.

**Comment by JStaWZL** [comment]
Sounds interesting. Sure, I can test that when it's ready.

**Comment by aaronzi** [comment]
Thank you very much. This is the updated image with the virtual scroller test: aaronzi/basyx-aas-ui:SNAPSHOT_08

**Comment by JStaWZL** [comment]
Hi again, I tried out the new image. Sadly it didn't fix the issue for us. I tried a few times but the web app seemed to just allocate to much memory what crashed my browser. There probably isn't any way around limiting the amount of data fetched into the application. 
I thought about the limitation you mentioned resulting from pagination of the results and the impact on our use case. It probably would be the case that the limitation of the search ability to the paginated result would be quiet a draw back. 
Thinking about the situation from my naive point of view I would come up with a idea like maybe fetch the AAS information in a paginated way like you described for showing those results on the side. At the same time maybe fetch from the shell-descriptors API for all shells paginated data for all AAS but for each result only keep the required information for searching, like idShort, id, ..., in an additional data structure. Discard all unneeded data. when the search is used, look up matches in the additional data structure and fetch the corresponding complete AAS information then and display it. One would also need pagination here in case results would be too much. If this would work, there would be ways to make things more optimized. But in any case I would discard ideas like this myself quickly because at some point this fails too and it never feels right to do crazy logic and work with big data structures on the client to compensate for missing functionality the backend should ideally provide.
That's just my thoughts. I don't know the AAS API as good as you guys do. We just started to work with this ecosystem a few month ago and experienced quiet a learning curve. I think the pagination feature would still be nice, so that the web app works without crashing also in case of bigger data amounts. For a perfect solution additional backend/API functionality would probably be needed as I see it.

**Comment by FrankSchnicke** [comment]
Thanks a lot for sharing your thoughts. After some internal discussions, in our opinion the only way to support this in a scalable way would be to implement a query support on Registry level for the _idShort_. 
However, while the specification of the AAS Repository does support a query for AAS by _idShort_, the specification of the registry API does not contain this. It would be possible to query the AAS Repository (or multiple ones) directly. This, however, would imply that all repositories are already known and would bypass the registries which could have other unintended consequences.

In consequence, the only feasible way of implementing your use case would be to extend the Registry to support this query endpoint. If you would like to contribute such an extension, I will gladly merge it. Additionally, you can describe your use case in the [AAS API Specs](https://github.com/admin-shell-io/aas-specs-api) repository and ask for an extension of the specification itself to include the endpoint.

**Comment by aaronzi** [comment]
I improved my front-end implementation for the virtual scroller, making it more sophisticated.
This is the new image: aaronzi/basyx-aas-ui:SNAPSHOT_10

I tested the image against a registry with 10.000 AAS (all having the same content). It worked very well with that and even the search by idShort is still possible. I would appreciate it if you could test this again on your end @JStaWZL.

The container is configured a bit differently than before. I was working on something else in the meantime. Here is a snipped from a docker-compose file for the new image version:

```yaml
# AAS Web UI
  aas-web-ui:
    image: aaronzi/basyx-aas-ui:SNAPSHOT_10
    container_name: aas-ui
    ports:
      - "3000:3000"
    environment:
      AAS_DISCOVERY_PATH: "http://localhost:8084/lookup/shells"
      AAS_REGISTRY_PATH: "http://localhost:8082/shell-descriptors"
      SUBMODEL_REGISTRY_PATH: "http://localhost:8083/submodel-descriptors"
      AAS_REPO_PATH: "http://localhost:8081/shells"
      SUBMODEL_REPO_PATH: "http://localhost:8081/submodels"
      CD_REPO_PATH: "http://localhost:8081/concept-descriptions"
    restart: always
    depends_on:
      aas-env:
        condition: service_healthy
```

**Comment by JStaWZL** [comment]
Hi @aaronzi I tried out your new build. First I had a hard time even loading the page with our data. The initially showed but with time the browser tab's memory filled up. I was wondering for quiet some time what was going on. As it turned out this was a browser specific problem. I use brave browser, and brave for some reason tries to fetch data for all AAS preemptively or something. In Firefox of Chrome those network requests didn't happen and the page can load fine. Maybe it should be checked if Safari or Edge have similar issues. I haven't those available. So the page renders and scrolling works fine even in our case with almost 100.000 AAS inside. So, this is great! 

But the search functionality is where the problem kicks in. Searching is possible but the moment you type one later the app freezes, you sometimes get the browser notification that the page is slowing down the browser and so on. If you just wait for like a minute the app is back on again and the results are actually shown. I tried this in Firefox and Chrome to avoid having a browser specific problem like before.

This is probably where we reach the point where we have to think about @FrankSchnicke 's comment. Thank you too, for sharing your thoughts. We will discuss this in our team too.

**Comment by aaronzi** [comment]
Interesting to hear that. I'm using Brave and it works fine with it. Also, the search is almost instant without any freezing issues. But if you are using 10 times more AAS this could be the reason as well. I only tested with 10.000. I also tried Edge (works fine). I will test Safari later.

But as already mentioned, the only real solution to this problem would be what @FrankSchnicke suggested in his comment.
